File: com.observerpattern/python/Subject.py
from abc import ABC, abstractmethod

from Observer import Observer

class Subject(ABC):
    @abstractmethod
    def add_observer(self, observer: Observer) -> None:
        pass
    
    @abstractmethod
    def remove_observer(self, observer: Observer) -> None:
        pass
    
    @abstractmethod
    def notify_observers(self) -> None:
        pass

File: Observer/ts/Subject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeatherStation = void 0;
class WeatherStation {
    constructor() {
        this.observers = [];
    }
    addObserver(observer) {
        this.observers.push(observer);
    }
    removeObserver(observer) {
        this.observers = this.observers.filter(o => o !== observer);
    }
    notifyObservers() {
        this.observers.forEach(observer => observer.update(this.weather));
    }
    setWeather(weather) {
        this.weather = weather;
        this.notifyObservers();
    }
}
exports.WeatherStation = WeatherStation;
